drivers/audio/audio_dma.c

```diff
@@ -0,0 +1,607 @@
+                               const struct audio_caps_s *caps);
+static int audio_dma_start(struct audio_lowerhalf_s *dev);
+#ifndef CONFIG_AUDIO_EXCLUDE_STOP
+static int audio_dma_stop(struct audio_lowerhalf_s *dev);
+#endif
+#ifndef CONFIG_AUDIO_EXCLUDE_PAUSE_RESUME
+static int audio_dma_pause(struct audio_lowerhalf_s *dev);
+static int audio_dma_resume(struct audio_lowerhalf_s *dev);
+#endif
+static int audio_dma_reserve(struct audio_lowerhalf_s *dev);
+static int audio_dma_release(struct audio_lowerhalf_s *dev);
+#endif
+static int audio_dma_allocbuffer(struct audio_lowerhalf_s *dev,
+                                 struct audio_buf_desc_s *bufdesc);
+static int audio_dma_freebuffer(struct audio_lowerhalf_s *dev,
+                                struct audio_buf_desc_s *bufdesc);
+static int audio_dma_enqueuebuffer(struct audio_lowerhalf_s *dev,
+                                   struct ap_buffer_s *apb);
+static int audio_dma_ioctl(struct audio_lowerhalf_s *dev, int cmd,
+                           unsigned long arg);
+static void audio_dma_callback(struct dma_chan_s *chan, void *arg,
+                               ssize_t len);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct audio_ops_s g_audio_dma_ops =
+{
+  .getcaps = audio_dma_getcaps,
+  .configure = audio_dma_configure,
+  .shutdown = audio_dma_shutdown,
+  .start = audio_dma_start,
+#ifndef CONFIG_AUDIO_EXCLUDE_STOP
+  .stop = audio_dma_stop,
+#endif
+#ifndef CONFIG_AUDIO_EXCLUDE_PAUSE_RESUME
+  .pause = audio_dma_pause,
+  .resume = audio_dma_resume,
+#endif
+  .allocbuffer = audio_dma_allocbuffer,
+  .freebuffer = audio_dma_freebuffer,
+  .enqueuebuffer = audio_dma_enqueuebuffer,
+  .ioctl = audio_dma_ioctl,
+  .reserve = audio_dma_reserve,
+  .release = audio_dma_release,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static int audio_dma_getcaps(struct audio_lowerhalf_s *dev, int type,
+                             struct audio_caps_s *caps)
+{
+  struct audio_dma_s *audio_dma = (struct audio_dma_s *)dev;
+
+  /* Validate the structure */
+
+  DEBUGASSERT(caps && caps->ac_len >= sizeof(struct audio_caps_s));
+  audinfo("type=%d ac_type=%d\n", type, caps->ac_type);
+
+  /* Fill in the caller's structure based on requested info */
+
+  caps->ac_format.hw  = 0;
+  caps->ac_controls.w = 0;
+
+  switch (caps->ac_type)
+    {
+      /* Caller is querying for the types of units we support */
+
+      case AUDIO_TYPE_QUERY:
+
+        /* Provide our overall capabilities.  The interfacing software
+         * must then call us back for specific info for each capability.
+         */
+
+        caps->ac_channels = 2;       /* Stereo output */
```

Comment:
Isn't it possible to support more channels in the future? Maybe it shouldn't be hard-code here.

Reply:
this should be a config option

Reply:
i think we need add DMA_QUERY in common DMA module,  and return capabilities in specific dma driver.
how do you think about it?

Reply:
i think we need add DMA_QUERY in common DMA module,  and return capabilities in specific dma driver.
how do you think about it?

Reply:
it isn't good to ask dma driver how many audio channel it can support since dma channel is general resource and shared by many hardware module(e.g. pcm, i2c, uart and spi).
The simple method is added the capability argument to audio_dma_initialize.

Reply:
I got it , thanks